lib/ret/web_push_subscription.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Ret.WebPushSubscription do
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Ecto.Query
+  use Retry
+
+  alias Ret.{EncryptedField, Hub, Repo, Statix, WebPushSubscription}
+
+  @schema_prefix "ret0"
+  @primary_key {:web_push_subscription_id, :id, autogenerate: true}
+  @push_rate_limit_seconds 60
+
+  schema "web_push_subscriptions" do
+    field(:p256dh, :string)
+    field(:endpoint, :string)
+    field(:auth, EncryptedField)
+    field(:last_notified_at, :utc_datetime)
+
+    belongs_to(:hub, Hub, references: :hub_id)
+
+    timestamps()
+  end
+
+  def subscribe_to_hub(
+        %Hub{hub_id: hub_id} = hub,
+        %{"endpoint" => endpoint, "keys" => %{"p256dh" => p256dh, "auth" => auth}} = subscription
+      ) do
+    find_by_hub_id_and_subscription(hub_id, subscription) ||
+      %WebPushSubscription{}
+      |> changeset_for_new(hub, %{p256dh: p256dh, auth: auth, endpoint: endpoint})
+      |> Repo.insert!()
+  end
+
+  def unsubscribe_from_hub(
+        %Hub{hub_id: hub_id},
+        subscription
+      ) do
+    with %WebPushSubscription{} = web_push_subscription <- find_by_hub_id_and_subscription(hub_id, subscription) do
+      web_push_subscription |> Repo.delete!()
+    end
+  end
+
+  def maybe_send(%WebPushSubscription{endpoint: endpoint, p256dh: p256dh, auth: auth} = web_push_subscription, body) do
+    if may_send?(web_push_subscription) do
+      subscription = %{
+        endpoint: endpoint,
+        keys: %{p256dh: p256dh, auth: auth}
+      }
+
+      retry with: exp_backoff() |> randomize |> cap(5_000) |> expiry(10_000) do
+        case WebPushEncryption.send_web_push(body, subscription) do
+          {:ok, _response} -> :ok
+          _ -> :error
+        end
+      after
+        result -> result
+      else
+        error -> error
+      end
+
+      Statix.increment("ret.web_push.hub.sent", 1)
+      web_push_subscription |> changeset_for_notification_sent |> Repo.update!()
```

Comment:
ah yes I agree the metrics and last notified state should only be updated on success will fix. i'm not sure what you mean about `is_attempting_to_send` -- is the concern around concurrent joins happening in different rooms resulting in multiple sends within the rate limiting window? if that's what you meant it's a fair point, i hadn't considered that, though I am not sure if I want to introduce additional state to mitigate that atm since I expect that case to be very rare.